packages/cozy-doctypes/src/Folder.js

```diff
@@ -0,0 +1,105 @@
+const Application = require('./Application')
+const CozyFile = require('./File')
+
+/**
+ * Class representing the folder model.
+ * @extends CozyFile
+ */
+class CozyFolder extends CozyFile {
+  /**
+   * Create a folder with a reference to the given document
+   * @param  {String}  path     Folder path
+   * @param  {Object}  document
+   * @return {Object}  Folder document
+   */
+  static async createFolderWithReference(path, document) {
+    const collection = this.cozyClient.collection(CozyFile.doctype)
+    const dirId = await collection.ensureDirectoryExists(path)
+    await collection.addReferencesTo(document, [
+      {
+        _id: dirId
+      }
+    ])
+
+    const { data: dirInfos } = await collection.get(dirId)
+
+    return dirInfos
+  }
+
+  /**
+   * Returns a "Magic Folder", given its id
+   * @param  {String} id Magic Folder id. Could be:
+   * * `CozyFolder.refs.ADMINISTRATIVE`
+   * * `CozyFolder.refs.PHOTOS`
+   * * `CozyFolder.refs.PHOTOS_BACKUP`
+   * * `CozyFolder.refs.PHOTOS_UPLOAD`
+   * @param {String} path Default path to use if magic folder does not exist
+   * @return {Object} Folder document
+   */
+  static async ensureMagicFolder(id, path) {
+    const magicFolderDocument = {
+      _type: Application.doctype,
+      _id: id
+    }
+    const folders = await CozyFolder.getReferencedFolders(magicFolderDocument)
+    const existingMagicFolder = folders.length ? folders[0] : null
+
+    if (existingMagicFolder) return existingMagicFolder
+
+    const magicFoldersValues = Object.values(CozyFolder.refs)
+    if (!magicFoldersValues.includes(id)) {
```

Comment:
```suggestion
    if (!CozyFolder.refs.hasOwnProperty(id)) {
```

This way, there is no need for the intermediate list. Inclusion check is better done with hashmaps than with list in general.

Reply:
I could be wrong, but here we are testing values and not keys, so I think `hasOwnProperty` is not appropriate are id is not `'ADMINISTRATIVE'` for example but `CozyFolder.refs.ADMINISTRATIVE` which value is `io.cozy.apps/administrative`.

Reply:
Oups, my bad !